File: src/frame.cpp
#include "frame.h"
#include "util.h"
#include <math.h> 
#include <algorithm>
#include <iomanip>


void Frame::load_cfg(const json & cfg) {
	max_theta = cfg["max_theta"].get<float>();
	min_dist = cfg["min_dist"].get<float>();
	inv_theta_grid_size = 1.0 / cfg["theta_grid_size"].get<float>();
	inv_dist_grid_size = 1.0 / cfg["dist_grid_size"].get<float>();

	lidar_height = cfg["lidar_height"].get<float>();
	max_slope = cfg["max_slope"].get<float>();
	dz_local = cfg["dz_local"].get<float>();
	dz_global = cfg["dz_global"].get<float>();

	n_angular_grids = static_cast<size_t>(2 * max_theta * inv_theta_grid_size);
	n_dist_grids = cfg["n_dist_grids"].get<size_t>();
	min_grd_pts = cfg["min_grd_pts"].get<size_t>();

}

void Frame::load_pts(const std::string& npz_filename) {
	// Load npz file
	cnpy::npz_t npz = cnpy::npz_load(npz_filename);
	size_t firings = npz["x"].shape[0];

	// Create pointers:
	float* xs = npz["x"].data<float>();
	float* ys = npz["y"].data<float>();
	float* zs = npz["z"].data<float>();
	float* rs = npz["r"].data<float>();
	bool * is_invalid = npz["invalid"].data<bool>();

	n_points = 0;
	float dxy, dz_cut, h_est;
	grd_candidates.reserve(max_points);

	for (size_t i = 0; i < firings; ++i) {
		if (n_points >= max_points) break;
		if (is_invalid[i]) continue;

		// Skip points that are below the global cut
		dxy = sqrt(xs[i]*xs[i] + ys[i]*ys[i]);
		dz_cut = std::max(std::min(dxy * max_slope, dz_global), dz_local);
		h_est = zs[i] + lidar_height;
		if (h_est + dz_cut < 0) continue;
		if (h_est < dz_cut)	grd_candidates.emplace_back(n_points);

		pos(n_points, 0) = xs[i];
		pos(n_points, 1) = ys[i];
		pos(n_points, 2) = zs[i];
		intensities[n_points] = rs[i];

		distXY[n_points] = dxy;
		thetas[n_points] = asin(xs[i] / dxy);

		++n_points;
	}
}


void Frame::gridding() {
	size_t grid_idx, ix, iy;
	float dz_cut;

	// Partition the points into the grids
	grid_indices.resize(n_dist_grids * n_angular_grids);

	// Loop through the points to get indices:
	for (size_t i = 0; i < n_points; ++i) {
		// Calculate the index in the for loop
		ix = static_cast<size_t>(std::min(std::max(thetas[i] + max_theta, 0.f) * inv_theta_grid_size, n_angular_grids-1.f));
		iy = static_cast<size_t>(std::min(std::max(distXY[i] - min_dist, 0.f) * inv_dist_grid_size, n_dist_grids-1.f));
		grid_idx = ix + iy * n_angular_grids;

		// Convert to the grid index
		grid_indices[grid_idx].emplace_back(i);
	}
}

void Frame::ground_detection() {
	//Reset the flags array
	flags.setZero();

	std::cout << "Ground candidates: " << grd_candidates.size() << std::endl;

	// Get grd_candidates
	std::array<float, 3> grd_par = plane_fit(pos(grd_candidates, Eigen::all));
	std::cout << std::fixed << std::setprecision(4);
	std::cout << "a=" << grd_par[0] << ", " << "b=" << grd_par[1] << ", " << "c=" << grd_par[2] << std::endl;

	// Get the default h_arr
	h_arr = pos.col(2) - grd_par[0] * pos.col(0) - grd_par[1] * pos.col(1) - grd_par[2];

	std::cout << "Seed points: " << ((h_arr < dz_local).cast<int>()).sum() << std::endl;
	size_t ix, iy, nb_idx;
	int g_ix, g_iy;

	// Loop through the grounds to get the local points marked
	for (size_t i = 0; i< grid_indices.size();  ++i) {
		// Skip this grid if not enough seed ground points
		// if (grid_indices[i].size() < min_grd_pts) continue;

		if ((h_arr(grid_indices[i]) < dz_local).cast<size_t>().sum() < min_grd_pts) continue;

		// Get the ix and iy
		ix = i % n_angular_grids;
		iy = i / n_angular_grids;

		// get Neiboring indices of the grid and perform the fit
		std::vector<size_t> nearby_to_fit_idx;

		for (int d_ix = -1; d_ix < 2; d_ix++) {
			for (int d_iy = -1; d_iy < 2; d_iy++) {
				// if (d_ix == 0 && d_iy==0) continue;
				g_ix = d_ix + ix;
				if (g_ix < 0 || g_ix >= n_angular_grids) continue;

				g_iy = d_iy + iy;
				if (g_iy < 0 || g_iy >= n_dist_grids) continue;

				nb_idx = g_ix + g_iy * n_angular_grids;

				for (auto idx : grid_indices[nb_idx]) {
					if (h_arr[idx] < dz_local) nearby_to_fit_idx.emplace_back(idx);
				}
			}
		}

		// perform the fit
		std::array<float, 3> local_grd_par = plane_fit(pos(nearby_to_fit_idx, Eigen::all));

		// Update the height for the points in the grid
		h_arr(grid_indices[i]) = pos(grid_indices[i], 2) - local_grd_par[0] * pos(grid_indices[i], 0)
		- local_grd_par[1] * pos(grid_indices[i], 1) - local_grd_par[2];

		// Label the local ground points
		size_t local_grd_pts = 0;
		for (auto idx : grid_indices[i]) {
			if (h_arr[idx] < dz_local) {
				// Compiling time to convert the type
				flags[idx] |= static_cast<uint8_t>(Flag::IS_GROUND);
				local_grd_pts++;
			}
		}

		// std::cout << local_grd_pts << " local ground points:" << "a="<< local_grd_par[0] <<", b="<<local_grd_par[1] << std::endl;

	}

	std::cout << "ground points: " << ((flags > 0).cast<int>()).sum() << std::endl;
}

File: algo/frame.py
"""@package single_frame_reco
This module takes in a full frame with full coverage of the lidar view,
returns the flag of each point for later processing.
Appreximation is made based on the assumption that the frame length is in the order
of 0.1s so the scene is roughly static.

Author: zukai.wang@cepton.com
Date: 9/18/2021
Version: 1.0.0
"""
import numpy as np
import plotly.graph_objects as go
import util

_IS_NOISE  = np.uint8(1)
_IS_GROUND = np.uint8(2)
_IS_BKG    = np.uint8(4)       # points belonging to the static backgrounds
_IS_VEHCLE = np.uint8(8)
_IS_SIGN   = np.uint8(16)
_IS_PED    = np.uint8(32)
_IS_BIKER  = np.uint32(64)


class Frame():
	"""
	Single Frame Reconstruction Result.
	"""
	def __init__(self, pts, cfg):
		self.cfg = cfg
		self.n_points = (~pts['invalid']).sum()
		self.pos = np.c_[pts['x'], pts['y'], pts['z']][~pts['invalid']]
		self.intensities = pts['r'][~pts['invalid']]
		self.h_arr = np.zeros(self.n_points)
		self.distXY = np.sqrt((self.pos[:, :2]**2).sum(axis=1))
		self.flags = np.zeros(self.n_points, dtype=np.uint8)

		self.grd_par = np.array([0, 0, -cfg["lidar_height"]])

	def gridding(self):
		# Partition the points by the angular & distance grids
		self.theta = np.arcsin(self.pos[:, 0] / self.distXY) 
		self.x_idx = ((self.theta + self.cfg['max_theta']) / self.cfg["theta_grid_size"]).astype(np.int16)
		self.y_idx = ((self.distXY - self.cfg['min_dist']) / self.cfg["dist_grid_size"]).astype(np.int16)

	def ground_detection(self):
		self.flags = np.zeros(self.n_points, dtype=np.uint8)
		# The bool array to indicate if the point is a potential ground point
		grd_candidates = np.abs(self.pos[:, 2] + self.cfg["lidar_height"]) < (self.distXY * self.cfg["max_slope"]).clip(self.cfg["dz_local"], self.cfg["dz_global"])
		
		grd_candidates = np.where(grd_candidates)[0]
		# Fit a global ground plane: z = ax + by + c
		a, b, c = util.plane_fit(self.pos[grd_candidates])

		# Set default heights
		self.h_arr = self.pos[:, 2] - a * self.pos[:, 0] - b * self.pos[:, 1] - c
		
		# Mark the selected seed ground points
		seed_grd = self.h_arr[grd_candidates] < self.cfg["dz_local"]
		self.flags[grd_candidates[seed_grd]] |= _IS_GROUND

		# Loop through the grids to get local points marked
		for iy in range(self.cfg["n_dist_grids"]):
			in_y = self.y_idx == iy
			if in_y.sum() < self.cfg["min_grd_pts"]:
				continue
			nearby_y = np.abs(self.y_idx - iy) < 2
			for ix in range(self.cfg["n_angular_grids"]):
				nearby_x = np.abs(self.x_idx - ix) < 2
				nearby_idx = np.where(nearby_x & nearby_y)[0]

				to_fit_idx = nearby_idx[self.flags[nearby_idx] & _IS_GROUND > 0]

				if len(to_fit_idx) < self.cfg["min_grd_pts"]:
					continue

				# Fit the local piece of the ground
				grd_par = util.plane_fit(self.pos[to_fit_idx])
				
				# clip the slope
				grd_par[:2] = grd_par[:2].clip(-self.cfg["max_slope"], self.cfg["max_slope"])

				# Update the heights for the local points
				in_x = self.x_idx == ix
				local_idx = np.where(in_x & in_y)[0]
				self.h_arr[local_idx] = self.pos[local_idx, 2] - self.pos[local_idx, :2] @ grd_par[:2] - grd_par[2]

				# Label the local ground points
				on_local_plane = (self.h_arr[local_idx] > -self.cfg["dz_global"]) & (self.h_arr[local_idx] < self.cfg["dz_local"])
				self.flags[local_idx[~on_local_plane]] = 0
				self.flags[local_idx[on_local_plane]] |= _IS_GROUND


	def visualize(self):
		is_grd = self.flags & _IS_GROUND > 0
		data = [go.Scatter3d(x=self.pos[is_grd, 0], y=self.pos[is_grd, 1], z=self.pos[is_grd, 2],
		       mode='markers', name=f"{is_grd.sum()} points on ground", marker=dict(size=1, opacity=0.3, color='green'))]
		
		abv_grd = self.h_arr >= self.cfg["dz_local"]
		data.append(go.Scatter3d(x=self.pos[abv_grd, 0], y=self.pos[abv_grd, 1], z=self.pos[abv_grd, 2],
		       mode='markers', name=f"{abv_grd.sum()} points above ground", marker=dict(size=1, opacity=0.3, color='red')))

		fig = go.Figure(data=data)

		fig.update_layout(title=f"{self.n_points} points in the frame", scene=dict(aspectmode='data'), template="plotly_dark")
		fig.show()
		

	def static_reco(self):
		# (N, 1)
		t_arr = self.ts - self.anchor_time

		# At anchor time, the front direction is ey
		alpha = self.omega[0] * t_arr
		beta  = self.omega[1] * t_arr
		gamma = self.omega[2] * t_arr

		cos_alpha, sin_alpha = np.cos(alpha), np.sin(alpha)
		cos_beta, sin_beta = np.cos(beta), np.sin(beta)
		cos_gamma, sin_gamma = np.cos(gamma), np.sin(gamma)

		ez0 = -sin_beta
		ez1 = sin_alpha * cos_beta
		ez2 = cos_alpha * cos_beta

		ey0 = cos_beta * sin_gamma
		ey1 = sin_alpha * sin_beta * sin_gamma + cos_alpha * cos_gamma
		ey2 = cos_alpha * sin_beta * sin_gamma - sin_alpha * cos_gamma

		# 3. ex = ey cross ez
		ex0 = ey1 * ez2 - ey2 * ez1
		ex1 = ey2 * ez0 - ey0 * ez2
		ex2 = ey0 * ez1 - ey1 * ez0

		pos_shift = self.velocity*t_arr

		self.positions = (np.vstack((self.xs*ex0+self.ys*ey0+self.zs*ez0 + ey0 * pos_shift,
									 self.xs*ex1+self.ys*ey1+self.zs*ez1 + ey1 * pos_shift,
									 self.xs*ex2+self.ys*ey2+self.zs*ez2 + ey2 * pos_shift))).T

